src/query/service/src/sql/optimizer/heuristic/where_optimizer.rs

```diff
@@ -0,0 +1,154 @@
+use crate::sql::plans::Filter;
+use crate::sql::plans::LogicalGet;
+use crate::sql::plans::PatternPlan;
+use crate::sql::plans::Prewhere;
+use crate::sql::plans::RelOp;
+use crate::sql::plans::Scalar;
+use crate::sql::MetadataRef;
+
+pub struct WhereOptimizer {
+    metadata: MetadataRef,
+    pattern: SExpr,
+}
+
+impl WhereOptimizer {
+    pub fn new(metadata: MetadataRef) -> Self {
+        Self {
+            metadata,
+            pattern: SExpr::create_unary(
+                PatternPlan {
+                    plan_type: RelOp::Filter,
+                }
+                .into(),
+                SExpr::create_leaf(
+                    PatternPlan {
+                        plan_type: RelOp::LogicalGet,
+                    }
+                    .into(),
+                ),
+            ),
+        }
+    }
+
+    fn collect_columns_no_subqueries(expr: &Scalar, columns: &mut ColumnSet) {
+        match expr {
+            Scalar::BoundColumnRef(column) => {
+                columns.insert(column.column.index);
+            }
+            Scalar::AndExpr(and) => {
+                Self::collect_columns_no_subqueries(and.left.as_ref(), columns);
+                Self::collect_columns_no_subqueries(and.right.as_ref(), columns);
+            }
+            Scalar::OrExpr(or) => {
+                Self::collect_columns_no_subqueries(or.left.as_ref(), columns);
+                Self::collect_columns_no_subqueries(or.right.as_ref(), columns);
+            }
+            Scalar::ComparisonExpr(cmp) => {
+                Self::collect_columns_no_subqueries(cmp.left.as_ref(), columns);
+                Self::collect_columns_no_subqueries(cmp.right.as_ref(), columns);
+            }
+            Scalar::FunctionCall(func) => {
+                for arg in func.arguments.iter() {
+                    Self::collect_columns_no_subqueries(arg, columns);
+                }
+            }
+            Scalar::CastExpr(cast) => {
+                Self::collect_columns_no_subqueries(cast.argument.as_ref(), columns);
+            }
+            // constant and subqueries is not collected
+            // TBD: how about aggregate function. I think aggregate function will not appear hear.
+            _ => {}
+        }
+    }
+
+    // analyze if the expression can be moved to prewhere
+    fn analyze(&self, expr: &Scalar, columns_to_scan: &ColumnSet) -> (bool, ColumnSet) {
+        let mut columns = ColumnSet::new();
+
+        // columns in subqueries are not considered
+        Self::collect_columns_no_subqueries(expr, &mut columns);
+
+        // viable conditions:
+        // 1. Condition depend on some column. Constant expressions are not moved.
+        // 2. Do not move conditions involving all queried columns.
+        // 3. Only current table columns are considered. (This condition is always true in current Pattern (Filter -> LogicalGet)).
+        (
+            !columns.is_empty() && columns.len() < columns_to_scan.len(),
```

Comment:
> And if they are equal, we can also push prewhere down to the storage layer.

I think not pushing down is better, and it's what exactly ClickHouse does.

For my current design in Databend, this will reduce one step in the data source state machine.

Reply:
ok